src/translate.py

```diff
@@ -809,16 +811,16 @@ class Cast(Expression):
     def dependencies(self) -> List[Expression]:
         return [self.expr]
 
-    def use(self) -> None:
+    def use(self, stack_info: StackInfo) -> None:
         # Try to unify, to make stringification output better.
-        self.expr.type.unify(self.type)
-        super().use()
+        self.expr.type.unify(self.type, stack_info.typemap)
+        super().use(stack_info)
 
     def needed_for_store(self) -> bool:
         if not self.reinterpret:
             # int <-> float casts should be emitted even for stores.
             return True
-        if not self.expr.type.unify(self.type):
+        if not self.expr.type.is_unified(self.type):
```

Comment:
as an alternative we could have use() store a property that tells whether unification succeeded

Reply:
Oh that's a way better nicer, good call.